File: CS/Persons.cs
using System;
using System.ComponentModel;

namespace TypeConverterTest {
	/// <summary>
	/// Summary description for DataSource.
	/// </summary>
	public class Persons : System.Collections.CollectionBase, System.ComponentModel.ITypedList {
		public Persons() {
		}

		public void Add(Person person) {
			this.InnerList.Add(person);
		}
		public Person this[string name] {
			get {
				foreach(Person person in this.InnerList)
					if (person.Name == name)
						return person;
				return null;
			}
		}

		PropertyDescriptorCollection ITypedList.GetItemProperties(System.ComponentModel.PropertyDescriptor[] listAccessors) {		
			System.Collections.ArrayList descriptors = new System.Collections.ArrayList();
			descriptors.Add(new PropertyDescriptor("Name"));
			descriptors.Add(new PropertyDescriptor("Address"));
			descriptors.Add(new PropertyDescriptor("Address_City"));

			System.ComponentModel.PropertyDescriptor[] propertyDescriptors = new System.ComponentModel.PropertyDescriptor[descriptors.Count];
			descriptors.CopyTo(propertyDescriptors, 0);
			return new PropertyDescriptorCollection(propertyDescriptors);
		}

		string ITypedList.GetListName(System.ComponentModel.PropertyDescriptor[] listAccessors) {
			return "";
		}	
	}
}

File: VB/Persons.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel

Namespace TypeConverterTest
	''' <summary>
	''' Summary description for DataSource.
	''' </summary>
	Public Class Persons
		Inherits System.Collections.CollectionBase
		Implements System.ComponentModel.ITypedList
		Public Sub New()
		End Sub

		Public Sub Add(ByVal person As Person)
			Me.InnerList.Add(person)
		End Sub
		Default Public ReadOnly Property Item(ByVal name As String) As Person
			Get
				For Each person As Person In Me.InnerList
					If person.Name = name Then
						Return person
					End If
				Next person
				Return Nothing
			End Get
		End Property

		Private Function GetItemProperties(ByVal listAccessors() As System.ComponentModel.PropertyDescriptor) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
			Dim descriptors As New System.Collections.ArrayList()
			descriptors.Add(New PropertyDescriptor("Name"))
			descriptors.Add(New PropertyDescriptor("Address"))
			descriptors.Add(New PropertyDescriptor("Address_City"))

			Dim propertyDescriptors(descriptors.Count - 1) As System.ComponentModel.PropertyDescriptor
			descriptors.CopyTo(propertyDescriptors, 0)
			Return New PropertyDescriptorCollection(propertyDescriptors)
		End Function

		Private Function GetListName(ByVal listAccessors() As System.ComponentModel.PropertyDescriptor) As String Implements ITypedList.GetListName
			Return ""
		End Function
	End Class
End Namespace